README.md
# ubuntu-21.04-docker

Vagrant Box with Ubuntu 21.04 and Docker

## Base image

Used base image [elegoev/ubuntu-21.04](https://app.vagrantup.com/elegoev/boxes/ubuntu-21.04)

## Directory Description

| directory | description                                          |
|-----------|------------------------------------------------------|
| inspec    | inspec test profiles with controls                   |
| packer    | packer build, provisioner and post-processor scripts |
| test      | test environment for provision & inspec development  |

## Vagrant

### Vagrant Cloud

- [elegoev/ubuntu-21.04-docker](https://app.vagrantup.com/elegoev/boxes/ubuntu-21.04-docker)

### Vagrant Plugins

- [vagrant-disksize](https://github.com/sprotheroe/vagrant-disksize)
- [vagrant-hosts](https://github.com/oscar-stack/vagrant-hosts)
- [vagrant-secret](https://github.com/tcnksm/vagrant-secret)
- [vagrant-vbguest](https://github.com/dotless-de/vagrant-vbguest)
- [vagrant-serverspec](https://github.com/vvchik/vagrant-serverspec)
- [vagrant-vmware-esxi](https://github.com/josenk/vagrant-vmware-esxi)

### Vagrantfile

    Vagrant.configure("2") do |config|

      $basebox_name="ubuntu-21.04-docker-test"
      $basebox_hostname="ubuntu-1804-docker-test"
      $src_image_name="elegoev/ubuntu-21.04-docker"
      $vb_group_name="basebox-docker-test"

      config.vm.define "#{$basebox_name}" do |machine|
        machine.vm.box = "#{$src_image_name}"
    
        # define guest hostname
        machine.vm.hostname = "#{$basebox_hostname}"

        machine.vm.provider "virtualbox" do |vb|
          vb.name = $basebox_name
          vb.cpus = 1
          vb.customize ["modifyvm", :id, "--memory", "1024" ]
          vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
          vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
          vb.customize [
            "modifyvm", :id, "--uartmode1", "file",
            File.join(Dir.pwd, "ubuntu-hirsute-21.04-cloudimg-console.log")
          ]
          vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
          vb.customize ["modifyvm", :id, "--vram", 256 ]
        end

      end   

    end

## Versioning

Repository follows sematic versioning  [![semantic versioning](https://img.shields.io/badge/semver-2.0.0-green.svg)](http://semver.org)

## Changelog

For all notable changes see [CHANGELOG](https://github.com/elegoev/basebox-ubuntu-21.04-docker/blob/main/CHANGELOG.md)

## License

Licensed under The MIT License (MIT) - for the full copyright and license information, please view the [LICENSE](https://github.com/elegoev/basebox-ubuntu-21.04-docker/blob/main/LICENSE) file.

## Issue Reporting

Any and all feedback is welcome.  Please let me know of any issues you may find in the bug tracker on github. You can find it [here.](https://github.com/elegoev/basebox-ubuntu-21.04-docker/issues)


CHANGELOG.md
# Change Log

## 1.0.0

**Implemented enhancements:**

- Docker installation


test/provision/vagrant-up.ps1
# set build environment
../../../../basebox-ubuntu-21.04/set-build-env.ps1

# run vagrant
vagrant up


test/download/cleanup-vagrant.ps1
# delete vagrant directory
$VagrantDirName = '.\.vagrant'
if (Test-Path $VagrantDirName) {
  Remove-Item $VagrantDirName -Recurse
}

# delete log file
$LogFilename = '.\ubuntu-hirsute-21.04-cloudimg-console.log'
if (Test-Path $LogFilename) {
  Remove-Item $LogFilename
}





test/provision/cleanup-vagrant.ps1
# delete ubuntu log file
$LogFileName = '.\ubuntu-hirsute-21.04-cloudimg-console.log'
if (Test-Path $LogFileName) {
  Remove-Item $LogFileName
}

# delete application file
$AppFileName = '.\installed-application.md'
if (Test-Path $AppFileName) {
  Remove-Item $AppFileName
}

# delete vagrant directory
$VagrantDirName = '.\.vagrant'
if (Test-Path $VagrantDirName) {
  Remove-Item $VagrantDirName -Recurse
}




packer/provisioners/base-provisioning.sh
#!/bin/bash

application_file_path="/vagrant/installed-application.md"

sudo apt update
while :
do
  sudo apt-get -y install \
       apt-transport-https \
       ca-certificates \
       curl \
       gnupg-agent \
       software-properties-common

  retcode=$?
  if [ $retcode -eq 0 ]; then
    break
  fi
  echo "sleep 5s ..."
  sleep 5
done

# add Docker’s official GPG key
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -

# add docker repository
sudo add-apt-repository \
   "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
   $(lsb_release -cs) \
   stable"

# install docker
sudo apt-get update
while :
do
  sudo apt-get -y install docker-ce docker-ce-cli containerd.io
  retcode=$?
  if [ $retcode -eq 0 ]; then
    break
  fi
  echo "sleep 5s ..."
  sleep 5
done

# add vagrant user to docker group
sudo groupadd docker
sudo gpasswd -a vagrant docker

# get docker version
DOCKER_VERSION=$(sudo docker version --format '{{.Server.Version}}')
echo "# Installed application "  > $application_file_path
echo "***                     " >> $application_file_path
echo "> docker $DOCKER_VERSION" >> $application_file_path


packer/post-processors/vagrant-cloud-create-box.ps1
# create vagrant cloud box

# check vagrant cloud token for packer (PACKER_VAGRANTCLOUD_TOKEN)
if (-not (Test-Path env:PACKER_VAGRANTCLOUD_TOKEN)) { 
  Write-Host "Environment variable 'PACKER_VAGRANTCLOUD_TOKEN' not set" -ForegroundColor Red
} else {

  # set vagrant image name
  $IMAGE_NAME = $env:IMAGE_NAME

  # show used token
  # Write-Host "token = $env:PACKER_VAGRANTCLOUD_TOKEN"

  # check if login user
  vagrant cloud auth login --token $env:PACKER_VAGRANTCLOUD_TOKEN
  if($?) {
    
    # check cloud image
    # vagrant cloud search $IMAGE_NAME --short | Select-String -Pattern 'elegoev/ubuntu-21.04'
    Write-Host "image name = $IMAGE_NAME"
    vagrant cloud box create "elegoev/$IMAGE_NAME" --no-private

  } else {
    Write-Host "Wrong vagrant cloud token configured" -ForegroundColor Red
  }

}







inspec/baseline-ubuntu-21.04-docker/controls/check_docker.rb
# copyright: 2021, Urs Voegele

title "check docker"

# check docker package
control "docker-1.0" do                     # A unique ID for this control
  impact 1.0                                # The criticality, if this control fails.
  title "check if docker is installed"      # A human-readable title
  desc "check docker package"
  describe packages(/docker/) do           # The actual test
    its('statuses') { should cmp 'installed' } 
  end
end

# check docker service
control "docker-2.0" do                     # A unique ID for this control
  impact 1.0                                # The criticality, if this control fails.
  title "check if docker is running"        # A human-readable title
  desc "check docker service"
  describe service('docker') do
    it { should be_installed }
    it { should be_enabled }
    it { should be_running }
  end
end


packer/post-processors/create-git-tag.ps1
# create git tag

$src_image_version = $env:IMAGE_VERSION

$vagrant_file_template = '
Vagrant.configure("2") do |config|

  $basebox_name="elegoev-ubuntu-21.04-docker-test"
  $basebox_hostname="ubuntu-1804-docker-test"
  $src_image_name="elegoev/ubuntu-21.04-docker"
  $vb_group_name="basebox-docker-test"

  config.vm.define "#{$basebox_name}" do |machine|
    machine.vm.box = "#{$src_image_name}"
    machine.vm.box_version = "$image_version"
    
    # define guest hostname
    machine.vm.hostname = "#{$basebox_hostname}"

    machine.vm.provider "virtualbox" do |vb|
      vb.name = $basebox_name
      vb.cpus = 1
      vb.customize ["modifyvm", :id, "--memory", "1024" ]
      vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
      vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
      vb.customize ["modifyvm", :id, "--uartmode1", "disconnected" ]
      vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
      vb.customize ["modifyvm", :id, "--vram", 256 ]
    end

    machine.trigger.after :destroy do |trigger|
      trigger.info = "Cleanup environment"
      trigger.run = {path: "./cleanup-vagrant.ps1"}  
    end

  end   

end
'

# create Vagrantfile for download test
$vagrant_file_expanded = $vagrant_file_template.Replace('$image_version', "$src_image_version")
Set-Content -Path './test/download/Vagrantfile' -Value "$vagrant_file_expanded"

# create git tag
git add --all
git commit -m "new basebox version $src_image_version"
git push 
git tag $src_image_version
git push --tags





